service/common/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogHandler.java

```diff
@@ -762,7 +777,190 @@ public LoadTableResponse updateTable(
+                } catch (CommitFailedException e) {
+                  if (!rollbackCompaction) {
+                    throw new ValidationFailureException(e);
+                  }
+                  // Since snapshot has already been created at the client end.
+                  // Nothing much can be done, we can move this
+                  // to writer specific thing, but it would be cool if catalog does this for us.
+                  // Inspect that the requirements states that snapshot
+                  // ref needs to be asserted this usually means in the update section
+                  // it has addSnapshot and setSnapshotRef
+                  UpdateRequirement.AssertRefSnapshotID addSnapshot = null;
+                  int found = 0;
+                  for (UpdateRequirement requirement : request.requirements()) {
+                    // there should be only add snapshot request
+                    if (requirement instanceof UpdateRequirement.AssertRefSnapshotID) {
+                      ++found;
+                      addSnapshot = (UpdateRequirement.AssertRefSnapshotID) requirement;
+                    }
+                  }
+
+                  if (found != 1) {
+                    // TODO: handle this case, find min snapshot id, to rollback to give it creates
+                    // lineage
+                    // lets not complicate things rn
+                    throw new ValidationFailureException(e);
+                  }
+
+                  Long parentSnapshotId = addSnapshot.snapshotId();
+                  // so we will first check all the snapshots on the top of
+                  // base on which the snapshot we want to commit is of type REPLACE ops.
+                  Long parentToRollbackTo = base.currentSnapshot().snapshotId();
+                  List<MetadataUpdate> updateToRemoveSnapshot = new ArrayList<>();
+                  while (!Objects.equals(parentToRollbackTo, parentSnapshotId)) {
+                    Snapshot snap = ops.current().snapshot(parentToRollbackTo);
+                    if (!DataOperations.REPLACE.equals(snap.operation())) {
+                      break;
+                    }
+                    updateToRemoveSnapshot.add(
+                        new MetadataUpdate.RemoveSnapshot(snap.snapshotId()));
+                    parentToRollbackTo = snap.parentId();
+                  }
+
+                  MetadataUpdate.SetSnapshotRef ref = null;
+                  found = 0;
+                  // find the SetRefName snapshot update
+                  for (MetadataUpdate update : request.updates()) {
+                    if (update instanceof MetadataUpdate.SetSnapshotRef) {
+                      ++found;
+                      ref = (MetadataUpdate.SetSnapshotRef) update;
+                    }
+                  }
+
+                  if (found != 1 || !Objects.equals(parentToRollbackTo, parentSnapshotId)) {
```

Comment:
Is this also validating that we don't have another branch which is based off of the commit we are rollingback?

Reply:
Ack, extracted the cross branch ref, validation and also added an UT for this scenario,